File: demos/vamp/VampPlanning.py
import numpy as np
import random
import copy
import vamp
import vamp.pointcloud
from functools import partial
from fire import Fire
from ompl import base as ob
from ompl import geometric as og
from ompl import tools as ot
from vamp_state_space import VampMotionValidator, VampStateValidityChecker, VampStateSpace

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
from viser_visualizer.viser_visualizer import ViserVisualizer

import time
import subprocess
import os
import shutil

POINTS_PER_SPHERE = 300
# Starting configuration
a = [0., -0.785, 0., -2.356, 0., 1.571, 0.785]

# Goal configuration
b = [2.35, 1., 0., -0.8, 0, 2.5, 0.785]
# Problem specification: a list of sphere centers
problem = [
    [0.55, 0, 0.25],
    [0.35, 0.35, 0.25],
    [0, 0.55, 0.25],
    [-0.55, 0, 0.25],
    [-0.35, -0.35, 0.25],
    [0, -0.55, 0.25],
    [0.35, -0.35, 0.25],
    [0.35, 0.35, 0.8],
    [0, 0.55, 0.8],
    [-0.35, 0.35, 0.8],
    [-0.55, 0, 0.8],
    [-0.35, -0.35, 0.8],
    [0, -0.55, 0.8],
    [0.35, -0.35, 0.8],
    ]


def isStateValid(s:ob.RealVectorStateType, env:vamp.Environment, dimension: int) -> bool:
    config = s[0:dimension]
    return vamp.panda.validate(config, env)


def create_environment(
    variation: float = 0.01,
    radius: float = 0.2,
    use_point_cloud: bool = False,
    points_per_sphere: int = POINTS_PER_SPHERE,
    ):
    """Create a VAMP environment with either spheres or point cloud obstacles.

    Returns (env, spheres, point_cloud) where point_cloud is None for sphere mode.
    """
    spheres = [np.array(sphere) for sphere in problem]
    random.shuffle(spheres)
    spheres_copy = copy.deepcopy(spheres)

    for sphere in spheres_copy:
        sphere += np.random.uniform(low=-variation, high=variation, size=(3,))

    env = vamp.Environment()
    point_cloud = None

    if use_point_cloud:
        point_cloud = np.vstack(
            [
                vamp.pointcloud.sphere_sample_surface(center, radius, points_per_sphere, 0.0)
                for center in spheres_copy
            ]
        ).astype(np.float32)

        r_min, r_max = vamp.panda.min_max_radii()
        filtered_point_cloud_list, elapsed_ns = vamp.filter_pointcloud(point_cloud, r_min, 2, (0,0,0), (-2,-2,-2), (2,2,2), True)
        n_added = env.add_pointcloud(filtered_point_cloud_list, r_min, r_max, 0.01)
    else:
        for sphere in spheres_copy:
            env.add_sphere(vamp.Sphere(sphere, radius))

    return env, spheres, point_cloud


def planOnce(
    variation: float = 0.01,
    radius: float = 0.2,
    use_point_cloud: bool = False,
    points_per_sphere: int = POINTS_PER_SPHERE,
    visualize: bool = False
    ):

    env, spheres, point_cloud = create_environment(
        variation, radius, use_point_cloud, points_per_sphere)

    # Use VAMP's robot module to initialize state space and validations
    robot = vamp.panda
    dimension = robot.dimension()
    space = VampStateSpace(robot=robot)
    si = ob.SpaceInformation(space)

    motion_validator = VampMotionValidator(si=si, env=env, robot=robot)
    state_validity_checker = VampStateValidityChecker(si=si, env=env, robot=robot)

    # Select a planner
    planner = og.RRTConnect(si)

    # Set validators
    si.setMotionValidator(motion_validator)
    si.setStateValidityChecker(state_validity_checker)

    # Build SimpleSetup object
    ss = og.SimpleSetup(si)
    ss.setPlanner(planner)

    # Create start and goal states
    start = si.allocState()
    start[0:dimension] = a
    goal = si.allocState()
    goal[0:dimension] = b
    ss.setStartAndGoalStates(start, goal)

    # Solve
    planning_start = time.time()
    result = ss.solve(5.0)
    if result:
        path = ss.getSolutionPath()
        print(f"Planning time: {time.time() - planning_start}")
        print(f"Generated path with {path.getStateCount()} states")

        if visualize:
            visualizer = ViserVisualizer(robot_name="panda",
                                        robot_dimension=dimension,
                                        port=8080)

            if point_cloud is not None:
                pc_colors = np.tile([204, 102, 51], (point_cloud.shape[0], 1))
                visualizer.add_point_cloud(point_cloud, color=pc_colors)
            else:
                for sphere in spheres:
                    visualizer.add_sphere(sphere, radius, color=(0.8, 0.4, 0.2))

            path.interpolate(50)
            states = path.getStates()
            trajectory = np.array([list(state[0:dimension]) for state in states])
            visualizer.visualize_trajectory(trajectory)
            visualizer.play_until_key_pressed()
    else:
        print(f"No solution found (time: {time.time() - planning_start})")


def planBenchmark(
    variation: float = 0.01,
    radius: float = 0.2,
    use_point_cloud: bool = False,
    points_per_sphere: int = POINTS_PER_SPHERE,
    n_trials: int = 100
    ):

    env, spheres, point_cloud = create_environment(
        variation, radius, use_point_cloud, points_per_sphere)

    # Use VAMP's robot module to initialize state space and validations
    robot = vamp.panda
    dimension = robot.dimension()
    space = VampStateSpace(robot=robot)
    si = ob.SpaceInformation(space)

    motion_validator = VampMotionValidator(si=si, env=env, robot=robot)
    state_validity_checker = VampStateValidityChecker(si=si, env=env, robot=robot)

    # Set validators
    si.setMotionValidator(motion_validator)
    si.setStateValidityChecker(state_validity_checker)

    # Build SimpleSetup object
    ss = og.SimpleSetup(si)

    start = si.allocState()
    start[0:dimension] = a
    goal = si.allocState()
    goal[0:dimension] = b
    ss.setStartAndGoalStates(start, goal)

    # Create Benchmark
    env_label = "PointCloud" if use_point_cloud else "Sphere"
    benchmark = ot.Benchmark(ss, f"Vamp Cage Planning Benchmark ({env_label})")
    benchmark.addPlanner(og.RRTConnect(si))
    benchmark.addPlanner(og.RRT(si))
    benchmark.addPlanner(og.KPIECE1(si))
    benchmark.addPlanner(og.LBKPIECE1(si))

    # Setup Benchmark
    req = ot.Request()
    req.maxTime = 1.0
    req.maxMem = 100.0
    req.runCount = n_trials
    req.displayProgress = True
    benchmark.benchmark(req)

    log_file = "benchmark.log"
    benchmark.saveResultsToFile(log_file)
    db_path = "benchmark.db"
    ot.readBenchmarkLog(db_path, [log_file], moveitformat=False)
    print(f"Database saved to {db_path}")

    try:
        import plannerarena.app
        print(f"Displaying benchmark results from {db_path}")
        plannerarena.app.DATABASE = os.path.abspath(db_path)
        plannerarena.app.run()
    except ImportError:
        print("PlannerArena not found, please install it from https://github.com/ompl/plannerarena")

def main(
    variation: float = 0.01,
    benchmark: bool = False,
    use_point_cloud: bool = False,
    n_trials: int = 100,
    radius: float = 0.2,
    visualize: bool = False
    ):

    if benchmark:
        planBenchmark(variation, radius, use_point_cloud, n_trials=n_trials)
    else:
        planOnce(variation, radius, use_point_cloud, visualize=visualize)
   
        
if __name__ == "__main__":
    Fire(main)

File: demos/vamp/VAMPPlanning.cpp
// demos/vamp/VAMPPlanning.cpp
#include <iostream>
#include <vector>
#include <array>
#include <chrono>

#include <boost/program_options.hpp>

// OMPL headers
#include <ompl/base/PlannerTerminationCondition.h>
#include <ompl/base/ProblemDefinition.h>
#include <ompl/base/SpaceInformation.h>
#include <ompl/base/spaces/RealVectorStateSpace.h>
#include <ompl/geometric/SimpleSetup.h>

// Planners
#include <ompl/geometric/planners/rrt/RRTConnect.h>
#include <ompl/geometric/planners/rrt/RRT.h>
#include <ompl/geometric/planners/kpiece/KPIECE1.h>
#include <ompl/geometric/planners/kpiece/LBKPIECE1.h>

// Benchmark
#include <ompl/tools/benchmark/Benchmark.h>

// VAMP integration headers
#include <ompl/vamp/Utils.h>
#include <ompl/vamp/VampStateValidityChecker.h>
#include <ompl/vamp/VampMotionValidator.h>
#include <ompl/vamp/VampStateSpace.h>

#include <vamp/collision/factory.hh>
#include <vamp/robots/panda.hh>

namespace ob = ompl::base;
namespace og = ompl::geometric;
namespace ot = ompl::tools;
namespace po = boost::program_options;


using Robot = vamp::robots::Panda;
using Environment = vamp::collision::Environment<vamp::FloatVector<vamp::FloatVectorWidth>>;

// Sphere obstacles
std::vector<std::array<float, 3>> obstacles = {
    {0.55, 0, 0.25},
    {0.35, 0.35, 0.25},
    {0, 0.55, 0.25},
    {-0.55, 0, 0.25},
    {-0.35, -0.35, 0.25},
    {0, -0.55, 0.25},
    {0.35, -0.35, 0.25},
    {0.35, 0.35, 0.8},
    {0, 0.55, 0.8},
    {-0.35, 0.35, 0.8},
    {-0.55, 0, 0.8},
    {-0.35, -0.35, 0.8},
    {0, -0.55, 0.8},
    {0.35, -0.35, 0.8},
};

void planOnce(){
    // Build a simple sphere obstacle environment
    vamp::collision::Environment<float> env_float;
    
    
    
    constexpr float radius = 0.2f;
    for (const auto& obs : obstacles)
    {
        env_float.spheres.emplace_back(
            vamp::collision::factory::sphere::array(obs, radius));
    }
    env_float.sort();
    
    // Convert to vectorized environment for SIMD collision checking
    Environment env(env_float);

    // Create state space
    auto space = std::make_shared<ompl::vamp::VampStateSpace<Robot>>();

    // Prints state space bounds (joint limits)
    std::cout << "Robot bounds:" << std::endl;
    for (std::size_t i = 0; i < Robot::dimension; ++i){
        std::cout << i << ": " << space->getBounds().low[i] << " to " << space->getBounds().high[i] << std::endl;
    }

    // Create simple setup
    og::SimpleSetup ss(space);

    auto si = ss.getSpaceInformation();

    // Sets state validity checker and motion validator, using SIMD-accelerated methods from VAMP
    si->setStateValidityChecker(
        std::make_shared<ompl::vamp::VampStateValidityChecker<Robot>>(si, env));
    si->setMotionValidator(
        std::make_shared<ompl::vamp::VampMotionValidator<Robot>>(si, env));

    // Define start and goal configurations (Panda 7 DOF)
    ob::ScopedState<> start(space);
    ob::ScopedState<> goal(space);
    
    std::array<double, 7> start_config = {0., -0.785, 0., -2.356, 0., 1.571, 0.785};
    std::array<double, 7> goal_config = {2.35, 1., 0., -0.8, 0., 2.5, 0.785};
    
    for (std::size_t i = 0; i < Robot::dimension; ++i)
    {
        start[i] = start_config[i];
        goal[i] = goal_config[i];
    }

    // Create RRTConnect planner (optional, a default will be automatically chosen otherwise)  
    auto planner = std::make_shared<og::RRTConnect>(si);
    ss.setPlanner(planner);

    // Setup problem definition
    ss.setStartAndGoalStates(start, goal);

    // Solve with 5 second timeout
    auto start_time = std::chrono::steady_clock::now();
    ob::PlannerStatus status = ss.solve(ob::timedPlannerTerminationCondition(5.0));
    auto end_time = std::chrono::steady_clock::now();
    
    auto duration_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start_time).count();
    auto duration_ms = duration_ns / 1e6;
    std::cout << "Planning time: " << duration_ms << " ms" << std::endl;

    if (status == ob::PlannerStatus::EXACT_SOLUTION)
    {
        std::cout << "Found solution!" << std::endl;
        auto path = ss.getSolutionPath();
        std::cout << "Path has " << path.getStateCount() << " states" << std::endl;
        path.print(std::cout);
    }
    else
    {
        std::cout << "No solution found" << std::endl;
    }
}

void planBenchmark(int runCount = 100){
    // Build a simple sphere obstacle environment
    vamp::collision::Environment<float> env_float;
    
    constexpr float radius = 0.2f;
    for (const auto& obs : obstacles)
    {
        env_float.spheres.emplace_back(
            vamp::collision::factory::sphere::array(obs, radius));
    }
    env_float.sort();
    
    // Convert to vectorized environment for SIMD collision checking
    Environment env(env_float);

    // create state space
    auto space = std::make_shared<ompl::vamp::VampStateSpace<Robot>>();

    std::cout << "Robot bounds:" << std::endl;
    for (std::size_t i = 0; i < Robot::dimension; ++i){
        std::cout << i << ": " << space->getBounds().low[i] << " to " << space->getBounds().high[i] << std::endl;
    }

    // Create simple setup
    og::SimpleSetup ss(space);

    auto si = ss.getSpaceInformation();

    si->setStateValidityChecker(
        std::make_shared<ompl::vamp::VampStateValidityChecker<Robot>>(si, env));
    si->setMotionValidator(
        std::make_shared<ompl::vamp::VampMotionValidator<Robot>>(si, env));

    // Define start and goal configurations (Panda 7 DOF)
    ob::ScopedState<> start(space);
    ob::ScopedState<> goal(space);
    
    std::array<double, 7> start_config = {0., -0.785, 0., -2.356, 0., 1.571, 0.785};
    std::array<double, 7> goal_config = {2.35, 1., 0., -0.8, 0., 2.5, 0.785};
    
    for (std::size_t i = 0; i < Robot::dimension; ++i)
    {
        start[i] = start_config[i];
        goal[i] = goal_config[i];
    }

    // Setup problem definition
    ss.setStartAndGoalStates(start, goal);

    // setup benchmark
    double memoryLimit = 4096;
    double runtimeLimit = 5.0;
    ot::Benchmark b(ss, "VAMP_Cage_Planning");
    ot::Benchmark::Request request(runtimeLimit, memoryLimit, runCount);

    b.addPlanner(std::make_shared<og::RRTConnect>(si));
    b.addPlanner(std::make_shared<og::RRT>(si));
    b.addPlanner(std::make_shared<og::KPIECE1>(si));    
    b.addPlanner(std::make_shared<og::LBKPIECE1>(si));

    b.benchmark(request);
    b.saveResultsToFile("vamp_cage_planning_benchmark_cpp.log");

    // Use python script to transfer .log to .db
    std::cout << "Results saved to vamp_cage_planning_benchmark_cpp.log" << std::endl;
    std::cout << "Use python script 'ompl/scripts/ompl_benchmark_statistics.py' to transfer .log to .db" << std::endl;
    // ompl/scripts/ompl_benchmark_statistics.py vamp_cage_planning_benchmark_cpp.log -d vamp_cage_planning_benchmark_cpp.db

}

int main(int argc, char** argv)
{
    po::options_description desc("Options");

    int runCount = 0;
    desc.add_options()
        ("help", "show help message")
        ("benchmark", po::value<int>(&runCount)->default_value(0), "Benchmark Planners for this number of trials")
        ;
    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    if (vm.count("help"))
    {
        std::cout << desc << std::endl;
        return 0;
    }

    if (runCount != 0U)
    {
        std::cout << "Running benchmark with " << runCount << " trials." << std::endl;
        planBenchmark(runCount);
    }
    else 
    {
        std::cout << "Running single planning instance." << std::endl;
        planOnce();
    }

    return 0;
}
